Add Baseline regression notes

ddbeck opened the pull request:
This is a work in progress toward fixing https://github.com/web-platform-dx/web-features/issues/1971.

This structures the existing Baseline regression comments as notes. I'm looking for early feedback on whether this structure for notes would be useful.

ddbeck committed:
Add Baseline regression notes

ddbeck committed:
Merge branch 'main' into regression-notes

ddbeck committed:
Structure the Baseline value change

ddbeck committed:
Add regression note for `createimagebitmap`

ddbeck committed:
Add Baseline regression notes to the schema

ddbeck committed:
Add Baseline regression note data guidelines

ddbeck committed:
Revise regression notes to follow the new guidelines

ddbeck committed:
Merge branch 'main' into regression-notes

ddbeck commented on features/popover.yml:
```diff
@@ -2,12 +2,16 @@ name: Popover
-# References:
-# - https://github.com/mdn/browser-compat-data/issues/22927
-# - https://bugs.webkit.org/show_bug.cgi?id=267688
+# notes:
+#   - date: 2024-09-18
+#     category: baseline-regression
+#     old_baseline_value: low
+#     new_baseline_value: false
+#     message: >
+#       Safari on iOS has a bug that prevents light dismiss (tapping outside the element to close it).
+#     citations:
+#       - https://bugs.webkit.org/show_bug.cgi?id=267688
+#       - https://github.com/mdn/browser-compat-data/issues/22927
```
When I opened this PR, I had written this note. However, the feature has since progressed. It occurred to me that we should not show irrelevant notes, so I commented it out. This is a rather new idea, so I'd like to delete this entirely, if we decide this is the right way to go.

Other options considered: some sort of "historic" flag on notes that are no longer relevant, or filtering historic notes from the published package. I figured both of those added complexity that wouldn't be very useful to anyone, least of all developers.

ddbeck commented on types.ts:
```diff
@@ -47,6 +47,8 @@ export interface FeatureData {
     /** Whether developers are formally discouraged from using this feature */
     discouraged?: Discouraged;
+    /** Notes about this feature */
+    notes?: BaselineRegressionNote[];
```
Good question, @captainbrosset!

I think it's _very_ likely that we'll have other note types and notes may need to coexist. My inclination was to introduce one note type, then add more. But I recognize it looks a little odd _now_ because there really ought to be just one note at a time under the constraints presented here. But there's some likely follow up note types:

First, consider the mirror of a regression: a case where we've _prevented_ a feature from advancing (think of it as `"baseline-delay"` categeory or something). There are two open cases of this right now, see https://github.com/web-platform-dx/web-features/issues/3228. It's possible (though rare) that a feature could have regressed and be prevented from advancing.

Second, there are cases where we might choose to _not_ regress a feature, like https://github.com/web-platform-dx/web-features/issues/2957 (though it's still my preference to let it regress). It seems to me it might be wise for us to (publicly) acknowledge the bug, even if we do not believe it enough to regress the feature (something like `"known-incompatibility"`, though ). That said, this type of note would be more tenuous (we might consider attaching notes to supported version number, for instance).

I imagine there are still further note types (e.g., something like solicitations for input on potentially deprecating or removing a feature, for instance, as in the case of XSLT), but I haven't given them a ton of thought yet.

rviscomi reviewed (changes requested):
I've tested this in Timebase (along with a patch to fix the note messages) to give an idea of how this might be integrated in developer tools/resources:

<img width="985" height="891" alt="image" src="https://github.com/user-attachments/assets/b68888bb-b615-49e8-84b0-734cf3d0b641" />

I've left a few comments in places where things don't appear as expected

ddbeck committed:
Merge branch 'main' into regression-notes

ddbeck committed:
Fix description duplication from @rviscomi

Co-authored-by: Rick Viscomi <rviscomi@users.noreply.github.com>

ddbeck committed:
Consolidate Baseline values into one subschema

This turns `"high" | "low" | false` into a named definition that's
reused in a couple places.

ddbeck committed:
Drop stale regression notes

ddbeck committed:
Test stale note check

ddbeck committed:
Drop `new_baseline_value` from the schema

`old_baseline_value` is always worse than `new_baseline_value`. Since we only have three stops, this means that only the status pairs [high, low], [high, false], and [low, false] can be represented as a regression. But that also means there's only one situation where a regressed feature could advance but not back to the level of the status before the regression (from high to false, then false to low). But at that point, the feature has had some time to regress, so that regression note is likely stale anyway.

ddbeck committed:
Merge branch 'main' into regression-notes

ddbeck committed:
FIx missing punctuation

ddbeck committed:
Add missing docstring

ddbeck committed:
Remove spurious edits

ABOsalah866 commented on features/streams.yml:
```diff
@@ -2,12 +2,6 @@ name: Streams
 description: The streams API creates, composes, and consumes continuously generated data.
 spec: https://streams.spec.whatwg.org/
 group: streams
-# TODO: https://github.com/web-platform-dx/web-features/issues/1971
-# Status changed: https://github.com/web-platform-dx/web-features/pull/2358, https://github.com/web-platform-dx/web-features/pull/2491
-# 2024-12-19 — low → false — Regressed status to match Caniuse, which considers support beginning at BYOB shipping.
-# 2025-01-30 — false → high — Split BYOB into a separate "readable-byte-streams" feature. Linked that one to Caniuse.
-# References:
-# - https://caniuse.com/streams
```
https://github.com/mdn/browser-compat-data/issues/new/choose

ddbeck commented:
This came up on today's call. There's interest in being able to explain Baseline regressions (e.g., for webstatus.dev feature subscriptions), but I have nervousness about committing to this specific schema.

One idea that came of the call would be to turn this into an _unpublished_ internal data mechanism, then experiment with the schema in the web features mapping repo, to make it safer to experiment with the data under real-world conditions.

When I have some time, I'll sync this up with main and refactor this as an internal field.

rviscomi commented:
Capturing a few motivating use cases that would benefit from this change:

- webstatus.dev gives developers the ability to subscribe to updates on web features, including regressions. This change would give subscribers more context behind why the feature regressed, with links to learn more.
- When a feature regresses, we reopen a feature's issue in the developer-signals repo. This change would allow us to provide more context to developers who might otherwise think it was reopened in error.
- Modern Web Guidance recommends using fallbacks whenever possible if a feature doesn't meet a Baseline target. If a feature regresses, the notes would be helpful context to include in the guides for coding agents to explain why fallbacks are suddenly necessary again.
